### PR TITLE
Generator for ProtoBuf (proto3)

### PR DESCRIPTION
**Pull request by TomGneuss**
As by issue #475, we requested a new generator for ProtoBuf. With this pull-request we present a possible solution for that.
In contrast to other generators, it only requires the `structure` module.
Since ProtoBuf is only a configuration language, it does not support functionalities like verification or object-oriented patterns like _Visitor_.
Thus, the code-base is minimal because also the expected output is reduced compared to the normal Core SDK.

### CODE REVIEW

**Comment by mristin** [comment]
@TomGneuss thanks a lot! Before I dig deeper, would you mind creating a separate smaller pull request for the renaming pascal to snake case? I was at first quite confused why other generators were changed, and this might make exploration of the history more difficult down the line.

If you lack time, I can do it for you -- please let me know what is easier for you.

**Comment by TomGneuss** [comment]
I ran the code-formatting script. As in #478, the mypy part fails.

I also discovered an issue in the generated proto files regarding properties that are not concrete types but interfaces. Since proto3 does not support interfaces (and inheritance in general), i fixed it by using the "oneof" keyword or introduced a *choice-message where necessary.

File: aas_core_codegen/cpp/description.py [hunk 1]
```diff
@@ -1,4 +1,5 @@
 """Render descriptions to documentation comments."""
+
```
Comment (g1zzm0) [inline aas_core_codegen/cpp/description.py:2]:
```suggestion
```

File: aas_core_codegen/cpp/constants/_generate.py [hunk 2]
```diff
@@ -1,4 +1,5 @@
 """Generate C++ constants corresponding to the constants of the meta-model."""
+
```
Comment (g1zzm0) [inline aas_core_codegen/cpp/constants/_generate.py:2]:
```suggestion
```

File: aas_core_codegen/cpp/enhancing/__init__.py [hunk 3]
```diff
@@ -1,4 +1,5 @@
 """Generate C++ code for enhancing model classes."""
+
```
Comment (g1zzm0) [inline aas_core_codegen/cpp/enhancing/__init__.py:2]:
```suggestion
```

File: aas_core_codegen/cpp/main.py [hunk 4]
```diff
@@ -1,4 +1,5 @@
 """Generate C++ code to handle models based on the meta-model."""
+
```
Comment (g1zzm0) [inline aas_core_codegen/cpp/main.py:2]:
```suggestion
```

File: aas_core_codegen/cpp/optionaling.py [hunk 5]
```diff
@@ -1,4 +1,5 @@
 """Infer and retrieve whether a node in our AST denotes a ``common::optional``."""
+
```
Comment (g1zzm0) [inline aas_core_codegen/cpp/optionaling.py:2]:
```suggestion
```

File: aas_core_codegen/cpp/stringification/__init__.py [hunk 6]
```diff
@@ -1,4 +1,5 @@
 """Generate C++ code to de/stringify enumerations and primitives."""
+
```
Comment (g1zzm0) [inline aas_core_codegen/cpp/stringification/__init__.py:2]:
```suggestion
```

File: aas_core_codegen/cpp/structure/_generate.py [hunk 7]
```diff
@@ -1,4 +1,5 @@
 """Generate the C++ data structures from the intermediate representation."""
+
```
Comment (g1zzm0) [inline aas_core_codegen/cpp/structure/_generate.py:2]:
```suggestion
```

**Review by g1zzm0 — CHANGES_REQUESTED** [review]

File: aas_core_codegen/typescript/transpilation.py [hunk 8]
```diff
@@ -1,4 +1,5 @@
 """Transpile meta-model Python code to TypeScript code."""
+
```
Comment (g1zzm0) [inline aas_core_codegen/typescript/transpilation.py:2]:
```suggestion
```

File: aas_core_codegen/typescript/unrolling.py [hunk 9]
```diff
@@ -1,4 +1,5 @@
 """Provide code generation for unrolling recursive calls and iterations."""
+
```
Comment (g1zzm0) [inline aas_core_codegen/typescript/unrolling.py:2]:
```suggestion
```

File: aas_core_codegen/typescript/verification/_generate.py [hunk 10]
```diff
@@ -1,4 +1,5 @@
 """Generate the invariant verifiers from the intermediate representation."""
+
```
Comment (g1zzm0) [inline aas_core_codegen/typescript/verification/_generate.py:2]:
```suggestion
```

File: aas_core_codegen/xsd/naming.py [hunk 11]
```diff
@@ -5,6 +5,7 @@
 :py:mod:`aas_core_codegen.naming`, which are used with different generators,
 these identifiers are used only for the XSD.
 """
+
```
Comment (g1zzm0) [inline aas_core_codegen/xsd/naming.py:8]:
```suggestion
```

File: aas_core_codegen/yielding/linear.py [hunk 12]
```diff
@@ -1,4 +1,5 @@
 """Translate control flows to linear flows with goto-statements and co-routines."""
+
```
Comment (g1zzm0) [inline aas_core_codegen/yielding/linear.py:2]:
```suggestion
```

File: setup.py [hunk 13]
```diff
@@ -4,6 +4,7 @@
 https://packaging.python.org/en/latest/distributing.html
 https://github.com/pypa/sampleproject
 """
+
```
Comment (g1zzm0) [inline setup.py:7]:
```suggestion
```

**Review by g1zzm0 — CHANGES_REQUESTED** [review]
Remove unnecessary blanks and empty lines

**Review by g1zzm0 — APPROVED** [review]
lgtm

**Comment by mristin** [comment]
@g1zzm0 why did you remove the empty lines? Wouldn't `black` take care of that?

**Comment by g1zzm0** [comment]
> @g1zzm0 why did you remove the empty lines? Wouldn't `black` take care of that?

idk. Didn't work for us. Fixed minor bugs in the proto gen.

**Comment by mristin** [comment]
@TomGneuss please let me know when you sync the pull request with the main branch and when it is ready for review.

**Comment by mristin** [comment]
@g1zzm0 @TomGneuss please make sure that you create separate pull requests for all the non-protobuf related changes. Otherwise, it is very hard to follow the git history and also to review.

**Comment by mristin** [comment]
> > @g1zzm0 why did you remove the empty lines? Wouldn't `black` take care of that?
> 
> idk. Didn't work for us. Fixed minor bugs in the proto gen.

I see the changes now also in the web. Please revert them (the removal of blank lines). We intentionally put a blank line after the module docstring (and I think `black` does so to?) for readability. Let's leave the formatting as it is for now.

**Comment by mristin** [comment]
You can try in your virtual environment:
```
pip3 uninstall black
pip3 install -e .[dev]
pip3 list 
black --version
```

Maybe you have a system-wide installed black which eclipses the local one?

**Comment by g1zzm0** [comment]
> You can try in your virtual environment:
We used the virtual Env like described in: https://github.com/aas-core-works/aas-core-codegen/blob/main/CONTRIBUTING.rst#create-a-development-environment

**Comment by mristin** [comment]
> > You can try in your virtual environment:
> > We used the virtual Env like described in: https://github.com/aas-core-works/aas-core-codegen/blob/main/CONTRIBUTING.rst#create-a-development-environment

Have you also installed black in this step?
https://github.com/aas-core-works/aas-core-codegen/blob/main/CONTRIBUTING.rst#install-development-dependencies

What do you get in the activated virtual environment when you call:
```
black --version
```
?

**Comment by g1zzm0** [comment]
> Have you also installed black in this step? https://github.com/aas-core-works/aas-core-codegen/blob/main/CONTRIBUTING.rst#install-development-dependencies

Yes.

> What do you get in the activated virtual environment when you call:
> 
> ```
> black --version
> ```
> 
> ?
22.3.0

**Comment by mristin** [comment]
@g1zzm0 @TomGneuss can you give me some hint what changed here?

aas_core_codegen/intermediate/_translate.py

Only formatting, right?
